### PR TITLE
MSAL updates to support loading launching custom tabs in current task

### PR DESCRIPTION
**Pull request by shoatman**
- Added configuration setting for authorization in current task
- Added default configuration value (false) for the same setting
- Fixed mergedConfiguration bug (already fixed in Dev)
- Exported new activity to receive authorization result for the current task scenario
- Initialized LibraryConfiguration based on PCA configuration (As discussed over email... in next release global configuration should be split out from PCA configuration)
- Changed test app main activity task affinity to null to demonstrate the problem with the current default behavior
- Updated the android manifest validation to add the receive authorization activity to it (CurrentTaskBrowserTabActivity vs BrowserTabActivity)

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        if(urlParameters != null){
+            /*
+            if(urlParameters.containsKey("state")) {
+                mTaskIdResponseFor = AndroidTaskStateGenerator.getTaskFromState(decodeState(urlParameters.get("state")));
+            }else{
+                IllegalStateException ex = new IllegalStateException("Did not receive a task id in the authorization response.");
+                com.microsoft.identity.common.logging.Logger.error(TAG, "Did not receive task id in the authorization response", ex);
+                throw ex;
+            }
+             */
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:93]: Do we need to delete this code?

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        }
+    }
+
+    @Override
+    protected void onDestroy() {
+        LocalBroadcastManager.getInstance(this).unregisterReceiver(mCloseBroadcastReceiver);
+        super.onDestroy();
+    }
+
+}
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:141]: nit: new line ending

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java [hunk 3]
```diff
@@ -367,6 +373,10 @@ public void setPowerOptCheckEnabled(Boolean powerOptCheckEnabled) {
         this.powerOptCheckEnabled = powerOptCheckEnabled;
     }
 
+    public Boolean authorizationInCurrentTask() {
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:376]: Following the pattern for other booleans...

```suggestion
    public Boolean isAuthorizationInCurrentTask() {
```

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfigurationFactory.java [hunk 4]
```diff
@@ -91,6 +92,11 @@ private static PublicClientApplicationConfiguration initializeConfigurationInter
             config.validateConfiguration();
         }
 
+        if(config.authorizationInCurrentTask()){
+            LibraryConfiguration libraryConfiguration = LibraryConfiguration.builder().authorizationInCurrentTask((true)).build();
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfigurationFactory.java:96]: nit: final

**Review by rpdome — COMMENTED** [review]
Left some comments

File: msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java [hunk 5]
```diff
@@ -259,7 +260,8 @@ public static boolean hasCustomTabRedirectActivity(@NonNull final Context contex
         for (final ResolveInfo info : resolveInfoList) {
             final ActivityInfo activityInfo = info.activityInfo;
 
-            if (activityInfo.name.equals(BrowserTabActivity.class.getName())) {
+            if (activityInfo.name.equals(BrowserTabActivity.class.getName()) ||
+                    activityInfo.name.equals(CurrentTaskBrowserTabActivity.class.getName())) {
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java:264]: I guess we need to be looking for a specific one based on library configuration? based on this code it seems the check would succeed if filter for `BrowserTabActivity` is defined but filter for `CurrentTaskBrowserTabActivity` is not defined...but what if the user specified to use `authorizartionInCurrentTast`? I guess they are destined to fail later in the code? Why not fail here?

File: testapps/testapp/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -43,7 +43,9 @@
         <activity
             android:name="com.microsoft.identity.client.testapp.MainActivity"
             android:theme="@style/AppTheme.NoActionBar"
-            android:windowSoftInputMode="stateHidden">
+            android:windowSoftInputMode="stateHidden"
+            android:taskAffinity=""
+            >
```
Comment (shahzaibj) [inline testapps/testapp/src/main/AndroidManifest.xml:48]:
```suggestion
            android:taskAffinity="">
```

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+ *     BrowserTabActivity will be launched when matching the custom url scheme.
+ *     &lt;data android:scheme="msalclientid" android:host="auth" /&gt;
+ * &lt;/intent-filter&gt;
+ * </pre>
+ */
+public final class CurrentTaskBrowserTabActivity extends Activity {
+    private static final String TAG = CurrentTaskBrowserTabActivity.class.getSimpleName();
+    private static final int REDIRECT_RECEIVED_CODE = 2;
+    private BroadcastReceiver mCloseBroadcastReceiver;
+    //private int mTaskIdResponseFor;
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:75]: do we still need this?

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+/**
+ * MSAL activity class (needs to be public in order to be discoverable by the os) to get the browser redirect with auth code from authorize
+ * endpoint. This activity has to be exposed by "android:exported=true", and intent filter has to be declared in the
+ * manifest for the activity.
+ * <p>
+ * When the AuthorizationAgent is launched, and we're redirected back with the redirect
+ * uri (the redirect must be unique across apps on a device), the os will fire an intent with the redirect,
+ * and the CurrentTaskBrowserTabActivity will be launched.
+ *
+ * Only use this if you've configured MSAL to use currenntTaskForAuthorizationActivities
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:55]:
```suggestion
 * Only use this if you've configured MSAL to use authorization_in_current_task
```

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 1 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:93]: Yep

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        final String response = getIntent().getDataString();
+        final HashMap<String, String> urlParameters = StringUtil.isEmpty(response) ? null : StringExtensions.getUrlParameters(response);
+        if(urlParameters != null){
+            /*
```
[the remark this answers is not in the document]
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:85]: Yes.  I should remove.

File: msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java (hunk 5 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java:264]: Good catch.  I'll fix this.

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 7 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:75]: Nope.  I'll remove.

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 2 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:141]: ran formatter

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfigurationFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfigurationFactory.java:96]: Yep.  We can do that.

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java [hunk 10]
```diff
@@ -154,6 +157,9 @@
     @SerializedName(POWER_OPT_CHECK_FOR_NETWORK_REQUEST_ENABLED)
     private Boolean powerOptCheckEnabled;
 
+    @SerializedName(AUTHORIZATION_IN_CURRENT_TASK)
+    private Boolean authorizationInCurrentTask;
```
[the remark this answers is not in the document]
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:161]: done.

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:85]: Done.

File: testapps/testapp/src/main/AndroidManifest.xml (hunk 6 above)
Comment (shoatman) [reply testapps/testapp/src/main/AndroidManifest.xml:48]: I'm actually going to remove this.  We need to modify this for different automation runs and manual test passes.

File: msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java (hunk 8 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java:55]: Done.  Thanks!

File: msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java (hunk 5 above)
Comment (shoatman) [reply msal/src/main/java/com/microsoft/identity/client/internal/MsalUtils.java:264]: I fixed this here and above in checkIntentFilterAddedToAppManifestForBrokerFlow....

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplicationConfiguration.java:161]: @shoatman did you forget to push the new change? (Saw comments marked as done, but no new commit pushed so far)

**Comment by shoatman** [comment]
PR didn't get last commit for some reason.

**Pull request closed without merging**